### PR TITLE
Initial commit for dsc on deepmind control suite, reacher task, pixel…

### PR DESCRIPTION
**Pull request by ktan38**
… observation.

### CODE REVIEW

File: simple_rl/agents/func_approx/ddpg/DDPGAgentClass.py [hunk 1]
```diff
@@ -39,11 +39,20 @@ def __init__(self, state_size, action_size, seed, device, lr_actor=LRA, lr_criti
...
 
         self.noise = OrnsteinUhlenbeckActionNoise(mu=np.zeros(action_size))
-        self.actor = Actor(state_size, action_size, device=device)
-        self.critic = Critic(state_size, action_size, device=device)
 
-        self.target_actor = Actor(state_size, action_size, device=device)
-        self.target_critic = Critic(state_size, action_size, device=device)
+
+        if pixel_observation:
+            self.actor = ConvActor(4, state_size, action_size, device=device)
```
Comment (abagaria) [inline simple_rl/agents/func_approx/ddpg/DDPGAgentClass.py]: Make the `in_channels` an input to the DDPG class as `frame_stack` or something like that

File: simple_rl/agents/func_approx/ddpg/model.py [hunk 2]
```diff
@@ -89,6 +89,110 @@ def get_action(self, state):
...
+    def __init__(self, in_channels,  state_dim, action_dim, h1=HIDDEN_1, h2=HIDDEN_2, device=torch.device("cpu"), seed=0):
+        super(ConvCritic, self).__init__()
+        self.conv1 = nn.Conv2d(in_channels, 32, kernel_size=8, stride=4)
+        self.bn1 = nn.BatchNorm2d(32)
+        self.conv2 = nn.Conv2d(32, 32, kernel_size=4, stride=1)
+        self.bn2 = nn.BatchNorm2d(32)
+        self.conv3 = nn.Conv2d(32, 32, kernel_size=4, stride=1)
+        self.bn3 = nn.BatchNorm2d(32)
+
+        self.linear1 = nn.Linear(98*64+action_dim, 200)
```
Comment (abagaria) [inline simple_rl/agents/func_approx/ddpg/model.py]: Can we add a comment here showing how we got 98*64?

File: simple_rl/agents/func_approx/ddpg/replay_buffer.py [hunk 3]
```diff
@@ -18,8 +18,12 @@ def __init__(self, buffer_size=BUFFER_SIZE, name_buffer='', seed=0):
         np.random.seed(seed)
 
     def add(self, state, action, reward, next_state, terminal):
-        assert isinstance(state, np.ndarray) and isinstance(action, np.ndarray) and \
-               isinstance(reward, (int, float)) and isinstance(next_state, np.ndarray)
+        #assert isinstance(state, np.ndarray), "state"
```
Comment (abagaria) [inline simple_rl/agents/func_approx/ddpg/replay_buffer.py]: lets bring this assert statement back, we should only have `np.ndarray`'s inside the replay buffer!

File: simple_rl/agents/func_approx/dsc/OptionClass.py [hunk 4]
```diff
@@ -176,8 +177,8 @@ def initialize_with_global_ddpg(self):
 
 		# Not using off_policy_update() because we have numpy arrays not state objects here
 		for state, action, reward, next_state, done in self.global_solver.replay_buffer.memory:
-			if self.is_init_true(state):
-				if self.is_term_true(next_state):
+			if self.name == "global_option":
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/OptionClass.py]: Try to revert this to the old logic for initialize_with_global_ddpg()

File: simple_rl/agents/func_approx/dsc/OptionClass.py [hunk 5]
```diff
@@ -176,8 +177,8 @@ def initialize_with_global_ddpg(self):
...
+				if reward:
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/OptionClass.py]: Dont check for `reward` in this fashion, can screw us when the reward function is not 0 vs 1

File: simple_rl/agents/func_approx/dsc/OptionClass.py [hunk 6]
```diff
@@ -234,7 +235,7 @@ def add_initiation_experience(self, states):
 		if len(states) >= self.buffer_length:
 			segmented_states = segmented_states[-self.buffer_length:]
 		if self.pixel_observations:
-			examples = [np.array(segmented_state.features())[-1, :, :] for segmented_state in segmented_states]
+			examples = [(np.array(segmented_state.features())[-1, :, :]) for segmented_state in segmented_states]
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/OptionClass.py]: Can we add a convenience function inside the OptionClass that takes a state object and returns a numpy array which represents the input the initiation classifier.

Comment (abagaria) [reply simple_rl/agents/func_approx/dsc/OptionClass.py]: For example:

`def f(x, n):
    return x[-n, ...]`

Comment (abagaria) [reply simple_rl/agents/func_approx/dsc/OptionClass.py]: Also, maybe add a comment on why the tuple-ing was needed?

File: simple_rl/agents/func_approx/dsc/OptionClass.py [hunk 7]
```diff
@@ -260,7 +262,7 @@ def construct_image_feature_matrix(examples):
 
 	def construct_negative_data_for_one_class_classifier(self):
 		""" Randomly sample states from the global solver's replay buffer and assume they are negative examples. """
-		states = [np.array(experience.state)[-1, :, :] for experience in self.global_solver.replay_buffer.memory]
+		states = [np.array(experience[0])[-1, :, :] for experience in self.global_solver.replay_buffer.memory]
 		num_positive_examples = len(list(itertools.chain.from_iterable(self.positive_examples)))
 		sampled_states = random.sample(states, k=num_positive_examples)
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/OptionClass.py:275]: Note for the future: we dont want to sample the same number of negatives as positives. We want to add a scaling term to the loss function instead so as to not lose the correct topology of the initiation set classifier.

File: simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py [hunk 8]
```diff
@@ -16,7 +16,7 @@
 
 # Other imports.
 from simple_rl.mdp.StateClass import State
-from simple_rl.agents.func_approx.dsc.OptionClass import Option
+from OptionClass import Option
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py:19]: we shouldn't need this

File: simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py [hunk 9]
```diff
@@ -102,10 +103,10 @@ def __init__(self, mdp, max_steps, lr_actor, lr_critic, ddpg_batch_size, device,
 		# We use (double-deep) (intra-option) Q-learning to learn the Q-values of *options* at any queried state Q(s, o)
 		# We start with this DQN Agent only predicting Q-values for taking the global_option, but as we learn new
 		# options, this agent will predict Q-values for them as well
+		print("this is state space size" + str(self.mdp.state_space_size()))
 		self.agent_over_options = DQNAgent(self.mdp.state_space_size(), 1, trained_options=self.trained_options,
 										   seed=seed, lr=1e-4, name="GlobalDQN", eps_start=1.0, tensor_log=tensor_log,
-										   use_double_dqn=True, writer=self.writer, device=self.device,
-										   pixel_observation=True)
+										   use_double_dqn=True, writer=self.writer, device=self.device)
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py]: Here, can we say `pixel_observation=self.mdp.pixel_observation` or as an argument to the DSC class

File: simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py [hunk 10]
```diff
@@ -188,7 +189,7 @@ def get_init_q_value_for_new_option(self, newly_trained_option):
 		state_option_pairs = newly_trained_option.final_transitions
 		q_values = []
 		for state, option_idx in state_option_pairs:
-			q_value = global_solver.get_qvalue(np.array(state.features()), option_idx)
+			q_value = global_solver.get_qvalue(state.features(), option_idx)
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py]: Make sure that `state.features()` is still a `np.array`. I think it is a `LazyFrame` or `FrameStack` thing rn

File: simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py [hunk 11]
```diff
@@ -515,7 +501,7 @@ def create_log_dir(experiment_name):
 
 	if "reacher" in args.env.lower():
 		from simple_rl.tasks.dm_fixed_reacher.FixedReacherMDPClass import FixedReacherMDP
-		overall_mdp = FixedReacherMDP(seed=args.seed, difficulty=args.difficulty, render=args.render)
+		overall_mdp = FixedReacherMDP(seed=args.seed, difficulty="hard", render=args.render)
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py]: revert

File: simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py [hunk 12]
```diff
@@ -528,11 +514,16 @@ def create_log_dir(experiment_name):
...
+		from simple_rl.tasks.dmcontrol_reacher.DMControlReacherMDPClass import DMControlReacherMDP
+		overall_mdp = DMControlReacherMDP(seed=args.seed, render=args.render)
+		state_dim = 6
+		action_dim = 2
 	else:
 		from simple_rl.tasks.gym.GymMDPClass import GymMDP
-		overall_mdp = GymMDP(args.env, render=args.render, pixel_observation=True)
-		state_dim = overall_mdp.state_space_size()
-		action_dim = overall_mdp.action_space_size()
+		overall_mdp = GymMDP(args.env, render=args.render)
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/SkillChainingAgentClass.py]: revert

File: simple_rl/agents/func_approx/dsc/utils.py [hunk 13]
```diff
@@ -128,6 +128,27 @@ def make_meshgrid(x, y, h=.02):
 	xx, yy = np.meshgrid(np.arange(x_min, x_max, h),
 						 np.arange(y_min, y_max, h))
 	return xx, yy
+def plot_one_class_initiation_images(option, episode=None, experiment_name=""):
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/utils.py:131]: set `xlim()` and `ylim()` using the bounds in the DMControlClass

File: simple_rl/tasks/gym/wrappers.py [hunk 14]
```diff
@@ -203,21 +203,25 @@ def __init__(self, env, num_stack, lz4_compress=False):
...
+        #self.realframes = deque(maxlen=num_stack)
 
         low = np.repeat(self.observation_space.low[np.newaxis, ...], num_stack, axis=0)
         high = np.repeat(self.observation_space.high[np.newaxis, ...], num_stack, axis=0)
         self.observation_space = Box(low=low, high=high, dtype=self.observation_space.dtype)
 
     def _get_observation(self):
         assert len(self.frames) == self.num_stack, (len(self.frames), self.num_stack)
-        return LazyFrames(list(self.frames), self.lz4_compress)
+        return (LazyFrames(list(self.frames), self.lz4_compress))
```
Comment (abagaria) [inline simple_rl/tasks/gym/wrappers.py]: lets not do this, lets fix whatever breaks while keeping this constant

Comment (abagaria) [reply simple_rl/tasks/gym/wrappers.py]: this is *probably* prematurely evaluating the LazyFrame object

File: simple_rl/tasks/gym/wrappers.py [hunk 15]
```diff
@@ -203,21 +203,25 @@ def __init__(self, env, num_stack, lz4_compress=False):
         self.lz4_compress = lz4_compress
 
         self.frames = deque(maxlen=num_stack)
+        #self.realframes = deque(maxlen=num_stack)
```
Comment (abagaria) [inline simple_rl/tasks/gym/wrappers.py:206]: lets not modify the gym wrappers unless absolutely necessary.

File: simple_rl/tasks/gym/wrappers.py [hunk 16]
```diff
@@ -203,21 +203,25 @@ def __init__(self, env, num_stack, lz4_compress=False):
...
         self.observation_space = Box(low=low, high=high, dtype=self.observation_space.dtype)
 
     def _get_observation(self):
         assert len(self.frames) == self.num_stack, (len(self.frames), self.num_stack)
-        return LazyFrames(list(self.frames), self.lz4_compress)
+        return (LazyFrames(list(self.frames), self.lz4_compress))
 
     def step(self, action):
         observation, reward, done, info = self.env.step(action)
+        observation = np.divide(observation[0] + observation[1] + observation[2], 3.0)
```
Comment (abagaria) [inline simple_rl/tasks/gym/wrappers.py]: The grayscale operation should not happen inside this wrapper. Either we find a Grayscale gym wrapper, or write it as a clean function in this same class

**Review by abagaria — CHANGES_REQUESTED** [review]
This is looking good lets make the minor edits i have mentioned and make another PR

File: simple_rl/agents/func_approx/ddpg/replay_buffer.py (hunk 3 above)
Comment (ktan38) [reply simple_rl/agents/func_approx/ddpg/replay_buffer.py]: we want to store lazyframes in there, will put appropriate assert statement in here
